### PR TITLE
Adding a config and a crypto module to the API client.

### PR DESCRIPTION
**Pull request by kiddinn**
This PR adds few things

1. **API Client**: New config module that reads and stores configuration information for Timesketch.
2. **API Client**: New crypto module that reads and stores both OAUTH credentials and `timesketch` credentials for Timesketch.
3. **Import Client**: New CLI module to ask questions as well as changes to the import CLI tool to make use of the new crypto and config libraries in the API Client.
4. Updated documentation.

### CODE REVIEW

File: api_client/python/timesketch_api_client/config.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        return self._config.keys()
+
+    @property
+    def missing(self):
+        """Return a list of missing parameters."""
+        return self.get_missing_config()
+
+    def get_config(self, name):
+        """Returns a value for a given config.
+
```
Comment (berggren) [inline api_client/python/timesketch_api_client/config.py:85]: +args

File: api_client/python/timesketch_api_client/config.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+            username=self._config.get('username'),
+            password=self._config.get('password', ''),
+            verify=self._config.get('verify', True),
+            client_id=self._config.get('client_id', ''),
+            client_secret=self._config.get('client_secret', ''),
+            auth_mode=auth_mode,
+        )
+
+    def get_missing_config(self):
+        """Returns a list of configuration parameters that are missing."""
```
Comment (berggren) [inline api_client/python/timesketch_api_client/config.py:139]: +returns

File: api_client/python/timesketch_api_client/config.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        needed_set = self.CLIENT_NEEDED
+        auth_mode = self._config.get('auth_mode', '')
+        if auth_mode.startswith('oauth'):
+            needed_set = needed_set.union(self.OAUTH_CLIENT_NEEDED)
+
+        configured_set = set(self._config.keys())
+        return list(needed_set.difference(configured_set))
+
+    def has_config(self, name):
```
Comment (berggren) [inline api_client/python/timesketch_api_client/config.py:148]: +raises

File: api_client/python/timesketch_api_client/config.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            'client_id': self._config.get('client_id', ''),
+            'client_secret': self._config.get('client_secret', ''),
+            'auth_mode': self._config.get('auth_mode', 'timesketch')
+        }
+
+        with open(file_path, 'w') as fw:
+            config.write(fw)
+
+    def set_config(self, name, value):
+        """Saves a config.
```
Comment (berggren) [inline api_client/python/timesketch_api_client/config.py:251]: s/Saves/Set ?

File: api_client/python/timesketch_api_client/crypto.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+
+logger = logging.getLogger('crypto_client_api')
+
+
+class CredentialStorage:
+    """Class to store and retrieve stored credentials."""
+
+    # A default shared secret that will be used as part of the key.
+    SHARED_KEY = 'thetta er mjog leynt'
```
Comment (berggren) [inline api_client/python/timesketch_api_client/crypto.py:37]: what is this language?! ;)

Comment (berggren) [reply api_client/python/timesketch_api_client/crypto.py:37]: no umlauts?

File: api_client/python/timesketch_api_client/crypto.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            '_token_uri': getattr(cred_obj, '_token_uri', ''),
+            '_client_id': getattr(cred_obj, '_client_id', ''),
+            '_client_secret': getattr(cred_obj, '_client_secret', ''),
+        }
+        if cred_obj.expiry:
+            data['expiry'] = cred_obj.expiry.isoformat()
+        data_string = json.dumps(data)
+
+        letters = string.ascii_letters
+        random_string = ''.join(
```
Comment (berggren) [inline api_client/python/timesketch_api_client/crypto.py:114]: this is a seed key, right? make the name reflect this in that case

File: api_client/python/timesketch_api_client/crypto.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            random.choice(letters) for _ in range(self.RANDOM_KEY_LENGTH))
+        key = self._get_key(random_string)
+        crypto = fernet.Fernet(key)
+
+        with open(file_path, 'wb') as fw:
+            fw.write(bytes(random_string, 'utf-8'))
```
Comment (berggren) [inline api_client/python/timesketch_api_client/crypto.py:120]: so the seed is in the file? so we only get half the key length in practice?

File: importer_client/python/timesketch_import_client/cli.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+import click
+
+def ask_question(question, input_type, default=None):
+    """Presents the user with a prompt with a default return value and a type.
+
+    Args:
+        question (str): the text that the user will be prompted.
+        input_type (type): the type of the input data.
+        default (object): default value for the question, optional.
```
Comment (berggren) [inline importer_client/python/timesketch_import_client/cli.py:24]: +returns

File: importer_client/python/tools/timesketch_importer.py [hunk 9]
```diff
@@ -228,50 +176,57 @@ def upload_file(
...
+    assistant.load_config_file()
+    assistant.load_config_dict(vars(options))
 
-    conf_password = options.password or config_options.get('password', '')
-    conf_pwd_prompt = options.pwd_prompt or config_options.get(
-        'pwd_prompt', False)
-    if not conf_password and conf_pwd_prompt:
+    conf_password = ''
+    if options.pwd_prompt:
         conf_password = getpass.getpass('Type in the password: ')
```
Comment (berggren) [inline importer_client/python/tools/timesketch_importer.py:190]: Maybe add some text about what this password is?

File: api_client/python/timesketch_api_client/config.py (hunk 1 above)
Comment (kiddinn) [reply api_client/python/timesketch_api_client/config.py:85]: done

File: api_client/python/timesketch_api_client/config.py (hunk 2 above)
Comment (kiddinn) [reply api_client/python/timesketch_api_client/config.py:149]: done

File: api_client/python/timesketch_api_client/config.py (hunk 3 above)
Comment (kiddinn) [reply api_client/python/timesketch_api_client/config.py:158]: has_config doesn't raise anything....

File: api_client/python/timesketch_api_client/config.py (hunk 4 above)
Comment (kiddinn) [reply api_client/python/timesketch_api_client/config.py:261]: reworded

File: api_client/python/timesketch_api_client/crypto.py (hunk 5 above)
Comment (kiddinn) [reply api_client/python/timesketch_api_client/crypto.py:180]: no umlauts ;)

File: api_client/python/timesketch_api_client/crypto.py (hunk 6 above)
Comment (kiddinn) [reply api_client/python/timesketch_api_client/crypto.py:242]: done

File: api_client/python/timesketch_api_client/crypto.py (hunk 7 above)
Comment (kiddinn) [reply api_client/python/timesketch_api_client/crypto.py:250]: half the key length?

what you mean?

fernet is a symmetric encryption and it requires a 32 bit key...

there are other ways to generate this key, this is one

File: importer_client/python/timesketch_import_client/cli.py (hunk 8 above)
Comment (kiddinn) [reply importer_client/python/timesketch_import_client/cli.py:24]: done

File: importer_client/python/tools/timesketch_importer.py (hunk 9 above)
Comment (kiddinn) [reply importer_client/python/tools/timesketch_importer.py]: done

**Comment by kiddinn** [comment]
ok, changes made, back to you

File: api_client/python/timesketch_api_client/crypto.py (hunk 7 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/crypto.py]: You are concatenating the random string that is stored in the file itself with a password from the user. Instead, can we use a key derivation function like PBKDF2HMAC? That way you only need to store the salt somewhere convenient. Looking at the docs, this is actually the preferred way: https://cryptography.io/en/latest/fernet/#using-passwords-with-fernet

Comment (kiddinn) [reply api_client/python/timesketch_api_client/crypto.py]: I was just reading this article before.... 

done

**Comment by kiddinn** [comment]
@berggren PTAL

File: api_client/python/timesketch_api_client/config.py (hunk 3 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/config.py:165]: Sorry, my bad

**Pull request merged**